src/libmongoc/tests/test-mongoc-sdam.c

```diff
@@ -215,6 +216,250 @@ test_sdam_cb (bson_t *test)
    mongoc_client_destroy (client);
 }
 
+/* Initialize a test context to run one SDAM integration test file.
+ *
+ * Do not use json_test_ctx_init to initialize a context. It sends commands to
+ * check for sessions support. That interferes with failpoints set on isMaster.
+ * Plus, SDAM tests have other special requirements that make initializing the
+ * test context a two step process (explained in run_one_integration_test).
+ */
+static void
+sdam_json_test_ctx_init_config (json_test_ctx_t *ctx,
+                                const json_test_config_t *config)
+{
+   memset (ctx, 0, sizeof (*ctx));
+   ctx->config = config;
+   bson_init (&ctx->events);
+   ctx->acknowledged = true;
+   ctx->verbose = test_framework_getenv_bool ("MONGOC_TEST_MONITORING_VERBOSE");
+   bson_init (&ctx->lsids[0]);
+   bson_init (&ctx->lsids[1]);
+   bson_mutex_init (&ctx->mutex);
+}
+
+static void
+sdam_json_test_ctx_init_client (json_test_ctx_t *ctx, mongoc_client_t *client)
+{
+   const char *db_name;
+   const char *coll_name;
+
+   ctx->test_framework_uri = mongoc_uri_copy (client->uri);
+   db_name = bson_lookup_utf8 (ctx->config->scenario, "database_name");
+   coll_name = bson_lookup_utf8 (ctx->config->scenario, "collection_name");
+   ctx->client = client;
+   ctx->db = mongoc_client_get_database (client, db_name);
+   ctx->collection = mongoc_database_get_collection (ctx->db, coll_name);
+}
+
+static void
+sdam_json_test_ctx_cleanup (json_test_ctx_t *ctx)
+{
+   mongoc_collection_destroy (ctx->collection);
+   mongoc_database_destroy (ctx->db);
+   bson_destroy (&ctx->lsids[0]);
+   bson_destroy (&ctx->lsids[1]);
+   bson_destroy (&ctx->events);
+   mongoc_uri_destroy (ctx->test_framework_uri);
+   bson_destroy (ctx->sent_lsids[0]);
+   bson_destroy (ctx->sent_lsids[1]);
+   bson_mutex_destroy (&ctx->mutex);
+}
+
+static bool
+sdam_integration_operation_cb (json_test_ctx_t *ctx,
+                               const bson_t *test,
+                               const bson_t *operation)
+{
+   bson_t reply;
+   bool res;
+
+   res =
+      json_test_operation (ctx, test, operation, ctx->collection, NULL, &reply);
+
+   bson_destroy (&reply);
+
+   return res;
+}
+
+/* Try to get a completely clean slate by disabling failpoints on all servers.
+ */
+static void
+deactivate_failpoints_on_all_servers (mongoc_client_t *client)
```

Comment:
Is this compensating for the `configureFailPoint` operation handler not tracking exactly which fail points are set? May be related to my earlier comment in the other file.

As-is, I might worry that this would stop working if any test were to set a fail point other than `failCommand` (I don't believe any currently do that).

Reply:
Yup. Separate follow-up ticket works for me. I'll resolve this and you can drop a link to the CDRIVER ticket in the other thread (https://github.com/mongodb/mongo-c-driver/pull/648#discussion_r451223867) when you get a chance, and then resolve that accordingly.